services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/api_client/_base.py

```diff
@@ -0,0 +1,141 @@
+                # waits 1, 4, 8 seconds between retries and gives up
+                stop=stop_after_attempt(zelf._request_max_retries),
+                wait=wait_exponential(min=1),
+                retry=retry_if_exception_type(_RetryRequestError),
+                reraise=True,
+            ):
+                with attempt:
+                    try:
+                        logger.debug("Calling retry_on_errors")
+                        response: Response = await request_func(zelf, *args, **kwargs)
+                        return response
+                    except ConnectError as e:
+                        assert e._request
+                        logger.info(
+                            "[%s/%s]Retry. Unexpected ConnectError while requesting '%s %s': %s",
+                            attempt.retry_state.attempt_number,
+                            zelf._request_max_retries,
+                            e._request.method,
+                            e._request.url,
+                            f"{e=}",
+                        )
+                        raise _RetryRequestError(e) from e
+                    except TransportError as e:
+                        raise ClientTransportError from e
+        except _RetryRequestError as e:
+            # raise original exception
+            assert e.__cause__
+            raise e.__cause__
+
+    return request_wrapper
+
+
+def expect_status(expected_code: int):
+    """
+    NOTE: always apply after `retry_on_errors`
+
+    raises:
+    - `UnexpectedStatusError`
+    - `httpx.HTTPError`
+    """
+
+    def decorator(
+        request_func: Callable[..., Awaitable[Response]]
+    ) -> Callable[..., Awaitable[Response]]:
+        assert asyncio.iscoroutinefunction(request_func)
+
+        @functools.wraps(request_func)
+        async def request_wrapper(zelf: "BaseHThinClient", *args, **kwargs) -> Response:
+            logger.debug("Calling expect_status")
+
+            response = await request_func(zelf, *args, **kwargs)
+            if response.status_code != expected_code:
+                raise UnexpectedStatusError(response, expected_code)
+
+            return response
+
+        return request_wrapper
+
+    return decorator
+
+
+class BaseHThinClient:
+    SKIP_METHODS: set[str] = {"close"}
+
+    def __init__(
+        self,
+        *,
+        request_max_retries: int,
+        base_url: Optional[URLTypes] = None,
+        timeout: Optional[TimeoutTypes] = None,
+    ) -> None:
+        self._request_max_retries: int = request_max_retries
+
+        client_args: dict[str, Any] = {}
+        if base_url:
+            client_args["base_url"] = base_url
+        if timeout:
+            client_args["timeout"] = timeout
+        self._client: AsyncClient = AsyncClient(**client_args)
+
+        # ensure all user defined public methods return `httpx.Response``
+
+        pubic_methods = [
+            x
+            for x in inspect.getmembers(self, predicate=inspect.ismethod)
```

Comment:
MINOR: for monosilable variables there is kind of a non-written rule: e.g. ``i`` or ``n`` denote integers and ``x,y`` some sort of flots. Now, if here is a member, please use something like ``m`` or something that can help readability

Reply:
I thought the the conversion for python is to use `x` in list comprehensions in the lack of a better name. I've changed it to `t` for tuple, since it's used as such.